plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs

```diff
@@ -70,18 +70,47 @@ data BuiltinMeaning val cost =
 type HasMeaningIn uni val = (Typeable val, HasConstantIn uni val)
 
 -- | A type class for \"each function from a set of built-in functions has a 'BuiltinMeaning'\".
-class (Typeable uni, Typeable fun, Bounded fun, Enum fun, Ix fun) => ToBuiltinMeaning uni fun where
+class (Typeable uni, Typeable fun, Bounded fun, Enum fun, Ix fun, Default (BuiltinVersion fun)) => ToBuiltinMeaning uni fun where
     -- | The @cost@ part of 'BuiltinMeaning'.
     type CostingPart uni fun
 
+    -- | See Note [Versioned builtins]
+    data BuiltinVersion fun
+
     -- | Get the 'BuiltinMeaning' of a built-in function.
-    toBuiltinMeaning :: HasMeaningIn uni val => fun -> BuiltinMeaning val (CostingPart uni fun)
+    toBuiltinMeaning :: HasMeaningIn uni val => BuiltinVersion fun -> fun -> BuiltinMeaning val (CostingPart uni fun)
 
 -- | Get the type of a built-in function.
-typeOfBuiltinFunction :: forall uni fun. ToBuiltinMeaning uni fun => fun -> Type TyName uni ()
-typeOfBuiltinFunction fun = case toBuiltinMeaning @_ @_ @(Term TyName Name uni fun ()) fun of
+typeOfBuiltinFunction :: forall uni fun. ToBuiltinMeaning uni fun => BuiltinVersion fun -> fun -> Type TyName uni ()
+typeOfBuiltinFunction ver fun = case toBuiltinMeaning @_ @_ @(Term TyName Name uni fun ()) ver fun of
     BuiltinMeaning sch _ _ -> typeSchemeToType sch
 
+{- Note [Versioned builtins]
+The purpose of the "versioned builtins" feature is to provide multiple, different denotations (implementations)
+for the same builtin(s).
+An example use of this feature is for "fixing" the behaviour of `ConsByteString` builtin to throw an error instead
+of overflowing its first argument.
+
+One denotation from each builtin is grouped into a 'BuiltinVersion'. Each Plutus Language version is linked
+to a specific 'BuiltinVersion' (done by plutus-ledger-api); e.g. plutus-v1 and plutus-v2 are linked to 'DefaultFunV1',
+whereas plutus-v3 changes the set of denotations to 'DefaultFunV2' (thus fixing 'ConsByteString').
+
+Each 'BuiltinVersion' (grouping) can change the denotation of one or more builtins --- or none, but what's the point in that.
+
+This 'BuiltinVersion' is modelled as a datatype *associated* to the `fun`. This associated datatype is required to
+provide an instance of 'Default' for quality-of-life purpose; the `def`ault builtin version is expected to point to the builtin-version
+that the plutus-tx/plutus-ir compiler is currently targetting.
+
+Note that, (old) denotations of a 'BuiltinVersion' cannot be removed or deprecated, once published to the chain.
+
+The way that this feature is implemented buys us more than we actually need:
+not only it allows the change of the denotation of a builtin, but also its expected *type signature*,
+i.e. type of arguments/result as well as number of arguments.
```

Comment:
Probably worth having a comment on cost models. E.g., what if a function requires different number of cost model parameters depending on the version. We haven't decided on an approach, but we can mention here that it's an open question.